go/pkg/exportsyntax/export-netspoc-syntax.1.md
# export-netspoc-syntax 1 "" Netspoc "User Manual"

# NAME

export-netspoc-syntax - Write Netspoc configuration as JSON to STDOUT

# SYNOPSIS

export-netspoc-syntax [options] netspoc-data [TYPE:NAME|TYPE: ...]

# DESCRIPTION

export-netspoc-syntax writes selected toplevel definitions as JSON to Stdout.

If a typed name is given, this object is written.
If a type with empty name is given, all objects of this type are written.
If no argument is given, all toplevel definitions are written.

Each definition is written as JSON object with key value pairs.
Definitions are grouped by TYPE, even if only a single object is exported.

# OPTIONS

**-q**, **--quiet**
:   Flag is ignored

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

With this Netspoc configuration

    #START
    network:n1 = {
     ip = 10.1.1.0/24;
     host:h1 = { ip = 10.1.1.11; owner = o1; }
     host:h2 = { range = 10.1.1.12-10.1.1.23; }
    }
    router:r1 = {
     managed;
     model = ASA;
     interface:n1 = {ip = 10.1.1.1; virtual = {ip = 10.1.1.9;} hardware = n1;}
    }
    #END

`export-netspoc-syntax netspoc/`

would generate this output:

    {"network":[
     {"name":"network:n1",
      "ip": [ "10.1.1.0/24" ],
      "hosts": {
       "host:h1": {
        "ip": [ "10.1.1.11" ],
        "owner": [ "o1" ]
       },
       "host:h2": {
        "range": [ "10.1.1.12 - 10.1.1.23" ]
       }
      }
     }],
     "router":[
     {"name":"router:r1",
      "managed": null,
      "model":["ASA"],
      "interfaces":{
       "interface:n1":{
        "ip": [ "10.1.1.1" ],
        "hardware": [ "n1" ],
        "virtual": { "ip": [ "10.1.1.9" ] }
       }
      }
     }]
    }

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/exportsyntax/export-netspoc-syntax.go
package exportsyntax

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/ast"
	"github.com/hknutzen/Netspoc/go/pkg/filetree"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/hknutzen/Netspoc/go/pkg/parser"
	"github.com/spf13/pflag"
)

func Main(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] netspoc-data [TYPE:NAME|TYPE: ...]\n%s",
			d.Args[0], fs.FlagUsages())
	}
	fs.BoolP("quiet", "q", false, "Flag is ignored")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Argument processing
	args := fs.Args()
	if len(args) < 1 {
		fs.Usage()
		return 1
	}
	filter := make(map[string]bool)
	for _, name := range args[1:] {
		filter[name] = true
	}
	path := args[0]
	// Group definitions by type.
	definitions := make(map[string][]jsonMap)
	err := filetree.Walk(path, func(input *filetree.Context) error {
		source := []byte(input.Data)
		path := input.Path
		aF, err := parser.ParseFile(source, path, 0)
		if err != nil {
			return err
		}
		for _, node := range aF.Nodes {
			name := node.GetName()
			i := strings.Index(name, ":")
			if len(filter) == 0 || filter[name[:i+1]] || filter[name] {
				typ := name[:i]
				definitions[typ] = append(definitions[typ], convertToMap(node))
			}
		}
		return nil
	})
	if err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}
	enc := json.NewEncoder(d.Stdout)
	enc.SetEscapeHTML(false)
	enc.Encode(definitions)
	return 0
}

type jsonMap map[string]any

func convertToMap(n ast.Toplevel) jsonMap {
	m := make(jsonMap)
	m["name"] = n.GetName()
	if d := n.GetDescription(); d != nil {
		m["description"] = d.Text
	}
	if x, ok := n.(ast.ToplevelWithAttr); ok {
		mapAttributes(m, x.GetAttributes())
	}
	switch x := n.(type) {
	case *ast.Protocol:
		m["value"] = x.Value
	case *ast.Protocolgroup:
		m["value_list"] = valueStrings(x.ValueList)
	case *ast.TopList:
		m["elements"] = elementStrings(x.Elements)
	case *ast.Network:
		if x.Hosts != nil {
			m["hosts"] = mapAttributes(make(jsonMap), x.Hosts)
		}
	case *ast.Router:
		m["interfaces"] = mapAttributes(make(jsonMap), x.Interfaces)
	case *ast.Area:
		if x.Border != nil {
			m["border"] = elementStrings(x.Border.Elements)
		}
		if x.InclusiveBorder != nil {
			m["inclusive_border"] = elementStrings(x.InclusiveBorder.Elements)
		}
	case *ast.Service:
		m["user"] = elementStrings(x.User.Elements)
		m["rules"] = mapRules(x.Rules)
		if x.Foreach {
			m["foreach"] = true
		}
	}
	return m
}

func mapRules(l []*ast.Rule) []jsonMap {
	result := make([]jsonMap, len(l))
	for i, r := range l {
		m := make(jsonMap)
		act := "permit"
		if r.Deny {
			act = "deny"
		}
		m["action"] = act
		m["src"] = elementStrings(r.Src.Elements)
		m["dst"] = elementStrings(r.Dst.Elements)
		m["prt"] = valueStrings(r.Prt.ValueList)
		if r.Log != nil {
			m["log"] = valueStrings(r.Log.ValueList)
		}
		result[i] = m
	}
	return result
}

func mapAttributes(m jsonMap, l []*ast.Attribute) jsonMap {
	for _, a := range l {
		if a.ValueList != nil {
			m[a.Name] = valueStrings(a.ValueList)
		} else if a.ComplexValue != nil {
			m[a.Name] = mapAttributes(make(jsonMap), a.ComplexValue)
		} else {
			m[a.Name] = nil
		}
	}
	return m
}

func valueStrings(l []*ast.Value) []string {
	result := make([]string, len(l))
	for i, v := range l {
		result[i] = v.Value
	}
	return result
}

func elementStrings(l []ast.Element) []string {
	result := make([]string, len(l))
	for i, e := range l {
		result[i] = e.String()
	}
	return result
}
